docs: init Realtime authz docs (#22803)
supabase/supabase: apps/docs/content/guides/realtime/authorization.mdx

<Admonition type="caution">

Realtime Authorization is in public alpha. To use Authorization for your Realtime Channels, use `realtime.js` version `2.10.0` or later.

</Admonition>

## How it works

Realtime uses tables in your database's `realtime` schema to track users and Channel topics. By writing RLS policies against these tables, you can control the access users have to a Channel topic and features within a Channel topic.

The validation is done when the user connects. When their WebSocket connection is established, their permissions are calculated based on:

- The RLS policies on tables in your `realtime` schema
- The user information sent as part of their [Auth JWT](/docs/guides/auth/jwts)
- The request headers
- The Channel topic the user is trying to connect to

<Admonition type="caution">

Increased RLS complexity can impact database performance and connection time, leading to higher connection latency and decreased join rates.

</Admonition>

## Helper functions

You can use the following helper functions when writing RLS policies:

### `realtime.topic`

Returns the Channel topic the user is attempting to connect to.

```sql

> create policy "authenticated users can read all messages in room_1"

Reviewer: this used to be `"authenticated users can read all messages"`, I changed it to what I _think_ is a more precise description, feel free to change back if I'm wrong

on "realtime"."messages"
for select
to authenticated
using (
  (select realtime.topic()) = 'room_1'
);
```

## Examples

The following examples use this schema:

```sql
create table public.profiles (
  email text not null,
  user_id uuid references auth.users (id)
);

create table public.rooms_users (
  topic text references realtime.messages (topic),
  user_id uuid references auth.users (id),
  created_at timestamptz default current_timestamp
);

// Trigger to populate profiles table
create or replace function insert_user () returns trigger as
$$
  begin
    insert into public.profiles (user_id, email) values (new.id, new.email);
    return new;
  end;
$$ language plpgsql;

create or replace trigger "on_new_auth_create_profile"
after insert on auth.users for each row
execute function insert_user ();

> revoke execute on function insert_user () from public, anon, authenticated;

Reviewer: I added this, not 100% sure if it's necessary?

Reviewer 2: it ends up being more correct 👍

> grant execute on function insert_user () TO supabase_auth_admin;

Reviewer: ```suggestion
grant execute on function insert_user () to supabase_auth_admin;
```

oops I missed this when editing

Author: Fixed.

grant insert on table public.profiles to supabase_auth_admin;
```

### Broadcast

The `extension` field on the `realtime.messages` table records the message type. For Broadcast messages, the value of `realtime.messages.extension` is `broadcast`. You can check for this in your RLS policies.

#### Allow a user to join (and read) a Broadcast topic

To join a Broadcast Channel, a user must have at least one read or write permission on the Channel topic.

Here, we allow reads (`select`s) for users who are linked to the requested topic within the relationship table `public.room_users`:

```sql
create policy "authenticated users can join topic"
on "realtime"."messages"
for select
to authenticated
using (
exists (
    select
      ru.user_id
    from
      rooms_users ru
    where

>       ru.user_id = (select auth.uid())

Reviewer: I wrapped all of these helper function calls in sub-selects because we recommend that in Performance Advisor, if that doesn't apply here, feel free to revert

Reviewer 3: nice one, thanks @charislam!

      and ru.topic = (select realtime.topic())
      and realtime.messages.extension in ['broadcast']
  )
);
```

Then, to join a topic with RLS enabled, instantiate the Channel with the `private` option set to `true`.

```typescript
import { createClient } from 'npm:@supabase/supabase-js@2.38.5'
const url = 'https://<project_ref>.supabase.com'
const apikey = '<api_key>'

const client = createClient(url, apikey)

const channel = client.channel('room_1', {

>   config: { private: true },

Reviewer: If `private` isn't `true`, then does that mean RLS doesn't apply, and anyone can join + read + write to the Channel? I wasn't entirely sure about this, probably worth mentioning up at the topic (under How It Works maybe?)

Reviewer 2: exactly and by default we assume it's false at this point in time

Author: Added, wdyt?

Reviewer: Perfect!

Reviewer 3: good call out, @charislam.

Let's add a note that mentions if private isn't true then the channel is not going to be checking AuthZ.

Here, we allow writes (sends) for users who are linked to the requested topic within the relationship table `public.room_users`:

```sql
create policy "authenticated users can send broadcast messages in a room"
on "realtime"."messages"
for insert
to authenticated
with check (
  exists (
    select
      ru.user_id
    from
      rooms_users ru
    where
      ru.user_id = (select auth.uid())
      and ru.topic = (select realtime.topic())
      and realtime.messages.extension in ['broadcast']
  )
);
```

### Presence

The `extension` field on the `realtime.messages` table records the message type. For Presence messages, the value of `realtime.messages.extension` is `presence`. You can check for this in your RLS policies.

#### Allow users to listen to Presence messages on a Channel

Create a policy for `select` on `realtime.messages` where `realtime.messages.extension` is `presence`.

```sql
create policy "authenticated users can listen to presence changes in a room"
on "realtime"."messages"
for select
to authenticated
using (
  exists (
    select
      ru.user_id
    from
      rooms_users ru
    where
      ru.user_id = (select auth.uid())

>       and ru.topic = (select realtime.channel_name())

Reviewer: Is `realtime.channel_name()` another helper function that we need to document (alongside `realtime.topic`)? What is the difference between these two (since they both seem to relate to the `room_users.topic` column?

Reviewer 2: realtime.channel_name() no longer exists and was removed, realtime.topic() took it's place

Author: Fixed.

Authorize both Presence and Broadcast by including both extensions in the `where` filter.

#### Broadcast and Presence read

Authorize Presence and Broadcast read in one RLS policy.

```sql
create policy "authenticated user can listen to Broadcast and Presence messages in room"
on "realtime"."messages"
for select
to authenticated
using (
  exists (
    select
      ru.user_id
    from
      rooms_users ru
    where
      ru.user_id = (select auth.uid())
      and ru.topic = (select realtime.channel_name())
      and realtime.messages.extension in ['broadcast', 'presence']
  )
);
```

### Broadcast and Presence write

Authorize Presence and Broadcast write in one RLS policy.

```sql
create policy "authenticated user can send Broadcast and Presence messages in a room"
on "realtime"."messages"
for insert
to authenticated
with check (
  exists (
    select
      ru.user_id
    from
      rooms_users ru
    where
      ru.user_id = (select auth.uid())
      and ru.name = (select realtime.channel_name())
      and realtime.messages.extension in ['broadcast', 'presence']
  )
);
```

## Interaction with Postgres Changes

Realtime Postgres Changes are separate from Channel authorization.

> Postgres Changes and Realtime Channel authorization are backwards compatible with your production application.

Reviewer: This part might need a bit more explanation, I wasn't entirely sure what it meant. That Postgres Changes are protected by the RLS policies of the underlying tables, which is separate from Channel authorization? (that was purely my guess)

Author: Attempted to clarify.

---
id: 'authorization'
title: 'Realtime Authorization'
description: 'Authorization for Supabase Realtime'
sidebar_label: 'Authorization'
---

You can control user access to Realtime Channels using RLS policies. By applying RLS, you can:

- Control which users can broadcast to a Channel
- Control which users can receive broadcasts from a Channel
- Control which users can publish their presence to a Channel
- Control which users can receive messages about the presence of other users

<Admonition type="caution">

Realtime Authorization is in public alpha. To use Authorization for your Realtime Channels, use `realtime.js` version `2.10.0` or later.

</Admonition>

> ## How it works

Reviewer 3: I think we left out how it makes a transaction and rolls it back on `realtime.messages`

Author: Added.

> Realtime uses tables in your database's `realtime` schema to track users and Channel topics. By writing RLS policies against these tables, you can control the access users have to a Channel topic and features within a Channel topic.

Author: > We should be specific here:

Updated. wdty?

> Let's make sure we refer to them as clients and not users.

I had this thought too but these are kind of interchangeable imo. eg `auth.users` table.

> We're calling them extensions so let's drop the mention of "features"

Updated also

> Realtime uses tables in your database's `realtime` schema to track users and Channel topics. By writing RLS policies against these tables, you can control the access users have to a Channel topic and features within a Channel topic.

Reviewer 3: @chasers yep, reads better. thanks!

### `realtime.topic`

Returns the Channel topic the user is attempting to connect to.

```sql
create policy "authenticated users can read all messages in room_1"
on "realtime"."messages"
for select
to authenticated
using (
  (select realtime.topic()) = 'room_1'
);
```

## Examples

The following examples use this schema:

```sql
create table public.profiles (
  email text not null,
  user_id uuid references auth.users (id)
);

create table public.rooms_users (
  topic text references realtime.messages (topic),
  user_id uuid references auth.users (id),
  created_at timestamptz default current_timestamp
);

// Trigger to populate profiles table
create or replace function insert_user () returns trigger as
$$
  begin
    insert into public.profiles (user_id, email) values (new.id, new.email);
    return new;
  end;
$$ language plpgsql;

create or replace trigger "on_new_auth_create_profile"
after insert on auth.users for each row
execute function insert_user ();

revoke execute on function insert_user () from public, anon, authenticated;
grant execute on function insert_user () to supabase_auth_admin;
grant insert on table public.profiles to supabase_auth_admin;
```

### Broadcast

> The `extension` field on the `realtime.messages` table records the message type. For Broadcast messages, the value of `realtime.messages.extension` is `broadcast`. You can check for this in your RLS policies.

Reviewer 3: We need to update docs b/c we never expanded Broadcast and Presence to extensions.

Reviewer 3: Can we also spell out what the extensions are? Like mention there are two right now: broadcast, presence.

```sql
create policy "authenticated users can send broadcast messages in a room"
on "realtime"."messages"
for insert
to authenticated
with check (
  exists (
    select
      ru.user_id
    from
      rooms_users ru
    where
      ru.user_id = (select auth.uid())
      and ru.topic = (select realtime.topic())
      and realtime.messages.extension in ('broadcast')
  )
);
```

### Presence

The `extension` field on the `realtime.messages` table records the message type. For Presence messages, the value of `realtime.messages.extension` is `presence`. You can check for this in your RLS policies.

#### Allow users to listen to Presence messages on a Channel

Create a policy for `select` on `realtime.messages` where `realtime.messages.extension` is `presence`.

```sql
create policy "authenticated users can listen to presence changes in a room"
on "realtime"."messages"
for select
to authenticated
using (
  exists (
    select
      ru.user_id
    from
      rooms_users ru
    where
      ru.user_id = (select auth.uid())
      and ru.topic = (select realtime.topic())
      and realtime.messages.extension in ('presence')
  )
);
```

> #### Allow users to send Presence messages on a channel

Reviewer 3: ```suggestion
#### Allow users to track Presence messages on a channel
```

> To update the Presence status for a user create a policy for `insert` on `realtime.messages` where the value of `realtime.messages.extension` is `presence`.

Reviewer 3: @filipecabaco I know we don't have `update` as a policy but we should really consider figuring out the full mapping now.

once our users write policies it'll be harder to change this up

  )
);
```

### Presence and Broadcast

Authorize both Presence and Broadcast by including both extensions in the `where` filter.

#### Broadcast and Presence read

Authorize Presence and Broadcast read in one RLS policy.

```sql
create policy "authenticated user can listen to Broadcast and Presence messages in room"
on "realtime"."messages"
for select
to authenticated
using (
  exists (
    select
      ru.user_id
    from
      rooms_users ru
    where
      ru.user_id = (select auth.uid())
      and ru.topic = (select realtime.topic())
      and realtime.messages.extension in ('broadcast', 'presence')
  )
);
```

#### Broadcast and Presence write

Authorize Presence and Broadcast write in one RLS policy.

```sql
create policy "authenticated user can send Broadcast and Presence messages in a room"
on "realtime"."messages"
for insert
to authenticated
with check (
  exists (
    select
      ru.user_id
    from
      rooms_users ru
    where
      ru.user_id = (select auth.uid())
      and ru.name = (select realtime.topic())
      and realtime.messages.extension in ('broadcast', 'presence')
  )
);
```

## Interaction with Postgres Changes

> Realtime Postgres Changes are separate from Channel authorization. The `private` Channel option does not apply to Postgres Changes.

Reviewer 3: We should mention that this is only for now. This is subject to change soon.

> When using Postgres Changes with RLS, database records are sent only to clients who are allowed to read them based on your RLS policies.

Reviewer 3: We should specify which RLS.

> database records are sent only to clients who are allowed to read them based on your RLS policies.

e.g. based on RLS policies on your tables and not `realtime.messages`.

> Realtime Postgres Changes are separate from Channel authorization. The `private` Channel option does not apply to Postgres Changes.
> 
> When using Postgres Changes with RLS, database records are sent only to clients who are allowed to read them based you your RLS policies.

Reviewer: ```suggestion
Realtime Postgres Changes are separate from Channel authorization. The `private` Channel option does not apply to Postgres Changes.

When using Postgres Changes with RLS, database records are sent only to clients who are allowed to read them based on your RLS policies.
```

Awesome, I think this is much clearer now

## Updating RLS policies

> When you change an RLS policy these changes are picked up when a user connects to a Channel.

Reviewer 3: Unnecessary: 

```suggestion
```

When a user connects, Realtime queries your database and generates a representation of the permissions for a user for a Channel.

> Permissions are cached for the duration of the connection so we don't have to query your database for each Channel message.

Reviewer 3: Not the duration of the connection but as long as client is subscribed to channel.

Author: A client can "leave" a topic and not disconnect but when the server kills the Channel I'm pretty sure the connection is actually terminated. eg when one of the `RealtimeChannel` callbacks respond with `{:stop, :shutdown, socket}` the process is stopped.

> RLS policies are re-checked when access tokens are updated on the Channel. Our client libraries default to updating the access token every 30 seconds.

Reviewer 3: How well is this going to scale if there are tons of users all refreshing token around the same time in the case they join around the same time? Is there a way we can batch?

Author: It'll scale as well as their database can scale. And yes we could batch at some point.

> If a user Channel permissions have been updated these changes will be picked up when the access token is sent over the Channel from the client.

Reviewer 3: ```suggestion
If a user Channel permissions have been updated these changes will be picked up when the refresh token is sent over the Channel from the client.
```

Author: Technically it's another "access token" which sent over to replace the current one used. A "refresh token" is not a thing. It's just a new "access token".

> If a new access token is never received on the Channel users will be disconnected when the JWT expires.
> 
> Make sure to expire your JWTs frequently.

Reviewer 3: e.g.

```suggestion
If a new access token is never received on the Channel client will be disconnected only when the JWT expires so make sure to set short expiration windows for your JWTs.
```

Author: I think mine is more clear.

When Realtime generates a policy for a client it does a test query on the `realtime.messages` table and then rolls it back. Realtime does not store any messages in your `realtime.messages` table.

Using Realtime authorization involves two steps:

- In your database, create an RLS policy on `realtime.messages`
- In your client, instantiate the Realtime Channel with the `config` option `private: true`

<Admonition type="caution">

Increased RLS complexity can impact database performance and connection time, leading to higher connection latency and decreased join rates.

</Admonition>

## Helper functions

You can use the following helper functions when writing RLS policies:

### `realtime.topic`

Returns the Channel topic the user is attempting to connect to.

```sql
create policy "authenticated users can read all messages in room_1"
on "realtime"."messages"
for select
to authenticated
using (
  (select realtime.topic()) = 'room_1'
);
```

## Examples

The following examples use this schema:

```sql
create table public.profiles (
  email text not null,
  user_id uuid references auth.users (id)
);

create table public.rooms_users (

>   topic text references realtime.messages (topic),

Reviewer 3: @chasers `realtime.messages` isn't storing anything so can't do this reference.

can you update the example SQL to match https://github.com/supabase/supabase/tree/master/examples/realtime/nextjs-authorization-demo#create-tables?

Author: Sure

---
id: 'authorization'
title: 'Realtime Authorization'
description: 'Authorization for Supabase Realtime'
sidebar_label: 'Authorization'
---

> In this section we detail how to set up Realtime authorization using RLS policies.

Reviewer 2: ```suggestion
In this section we detail how to set up Realtime authorization using Row Level Security policies.
```

Author: Out of date

<Admonition type="note">

Realtime authorization is in public alpha.

</Admonition>

<Admonition type="note">

Use realtime.js version `2.10.0` or greater.

</Admonition>

## How it works

Realtime uses tables in the `realtime` schema to allow you to write RLS policies which will determine the access your users have to a Channel topic and features within a Channel topic.

The validation itself is done on connect per user. For each WebSocket connection we generate the policy for your user based on:

- the RLS policies on tables in your `realtime` schema with the connection context including the JWT
- request headers set on connect
- the Channel topic the user is trying to connect to

<Admonition type="warning">

Increased RLS complexity can impact database performance and connection time, leading to higher connection latency and decreased join rates.

</Admonition>

## Helper functions

We've included a helper function to facilitate building policies.

### realtime.topic

Returns the Channel topic which the user is attempting to connect to.

```sql
CREATE POLICY "authenticated users can read all messages on a topic"
ON "realtime"."messages"
FOR SELECT
TO authenticated
USING (

>   realtime.topic = 'room_1'

Reviewer 2: ```suggestion
  realtime.topic() = 'room_1'
```

);
```

## Authorize Channels

Channels serve as our WebSocket representation.

With Realtime authorization we can authorize features within Channel topics or limit joining topics altogether.

### Example schema

Using the following schema for our examples below.

```sql
CREATE TABLE public.profiles (
  email text NOT NULL,
  user_id uuid REFERENCES auth.users (id)
);

CREATE TABLE public.rooms_users (
  topic text REFERENCES realtime.messages (topic),
  user_id uuid REFERENCES auth.users (id),
  created_at timestamptz DEFAULT CURRENT_TIMESTAMP
);

// Trigger to populate profiles table
CREATE OR REPLACE FUNCTION insert_user () RETURNS TRIGGER AS
$$
  BEGIN
    INSERT INTO public.profiles (user_id, email) VALUES (NEW.id, NEW.email); RETURN NEW;
  END;
$$ LANGUAGE plpgsql;

CREATE OR REPLACE TRIGGER "on_new_auth_create_profile"
AFTER INSERT ON auth.users FOR EACH ROW
EXECUTE FUNCTION insert_user ();

> GRANT EXECUTE ON FUNCTION insert_user () TO supabase_auth_admin;

Reviewer 2: ```suggestion
GRANT EXECUTE ON FUNCTION insert_user () TO supabase_auth_admin;
GRANT INSERT ON TABLE public.profiles TO supabase_auth_admin;
```

const url = 'https://<project_ref>.supabase.com'
const apikey = '<api_key>'

const client = createClient(url, apikey)

const channel = client.channel('room_1', {
  config: { private: true },
})

channel
  .on('broadcast', { event: 'test' }, (payload) => console.log(payload))
  .subscribe((status: string, err: any) => {
    if (status === 'SUBSCRIBED') {
      console.log('Connected!')
    } else {
      console.error(err)
    }
  })
```

## Authorize Broadcast read and write

Broadcast is represented by the `extension` field on the `realtime.messages` table.

If a message is a Broadcast message the value of `realtime.messages.extension` will be `broadcast`.

### Broadcast read

To grant users permission to read Broadcast messages on a Channel topic create a policy for `SELECT` operations on `realtime.messages` where `extension` is `broadcast`.

```sql
CREATE POLICY "authenticated users can receive broadcast messages in a room"
ON "realtime"."messages"
FOR SELECT
TO authenticated
USING (
  EXISTS (
    SELECT
      ru.user_id
    FROM
      rooms_users ru
    WHERE
      ru.user_id = auth.uid()
      AND ru.topic = realtime.topic()

>       AND realtime.messages.extension in ['broadcast']

Reviewer 2: ```suggestion
      AND realtime.messages.extension in ('broadcast')
```

  )
);
```

### Broadcast write

To authorize sending Broadcast messages create a policy for `INSERT` where the value of `realtime.messages.extention` is `broadcast`.

```sql
CREATE POLICY "authenticated users can send broadcast messages in a room"
ON "realtime"."messages"
FOR INSERT
TO authenticated
WITH CHECK (
  EXISTS (
    SELECT
      ru.user_id
    FROM
      rooms_users ru
    WHERE
      ru.user_id = auth.uid()
      AND ru.topic = realtime.topic()

>       AND realtime.messages.extension in ['broadcast']

Reviewer 2: ```suggestion
      AND realtime.messages.extension in ('broadcast')
```

  )
);
```

## Authorize Presence read and write

> Presnece is also represented by the `extension` field on the `realtime.messages` table.

Reviewer 2: ```suggestion
Presence is also represented by the `extension` field on the `realtime.messages` table.
```

If a message is a Presence message the value of `realtime.messages.extension` will be `presence`.

### Presence read

Create a policy for `SELECT` on `realtime.messages` where `realtime.messages.extension` is `presence`.

```sql
CREATE POLICY "authenticated users can listen to presence changes in a room"
ON "realtime"."messages"
FOR SELECT
TO authenticated
USING (
  EXISTS (
    SELECT
      ru.user_id
    FROM
      rooms_users ru
    WHERE
      ru.user_id = auth.uid()
      AND ru.topic = realtime.channel_name()

>       AND realtime.messages.extension in ['presence']

Reviewer 2: ```suggestion
      AND realtime.messages.extension in ('presence')
```

  )
);
```

### Presence write

To update the Presence status for a user create a policy for `INSERT` on `realtime.messages` where the value of `realtime.messages.extension` is `presence`.

```sql
CREATE POLICY "authenticated users can register and update their presence in a room"
ON "realtime"."messages"
FOR INSERT
TO authenticated
WITH CHECK (
  EXISTS (
    SELECT
      ru.user_id
    FROM
      rooms_users ru
    WHERE
      ru.user_id = auth.uid()
      AND ru.name = realtime.channel_name()

>       AND realtime.messages.extension in ['presence']

Reviewer 2: ```suggestion
      AND realtime.messages.extension in ('presence')
```

  )
);
```

## Authorize Presence and Broadcast

> Authorize Presence and Broadcast in two RLS policies by simply including both extensions in our where filter.

Reviewer 2: ```suggestion
Authorize Presence and Broadcast in two RLS policies including both extensions in our where filter.
```

Author: Stale

### Broadcast and Presence read

Authorize Presence and Broadcast read in one RLS policy.

```sql
CREATE POLICY "authenticated user can listen to Broadcast and Presence messages in room"
ON "realtime"."messages"
FOR SELECT
TO authenticated
USING (
  EXISTS (
    SELECT
      ru.user_id
    FROM
      rooms_users ru
    WHERE
      ru.user_id = auth.uid()
      AND ru.topic = realtime.channel_name()

>       AND realtime.messages.extension in ['broadcast', 'presence']

Reviewer 2: ```suggestion
      AND realtime.messages.extension in ('broadcast', 'presence')
```

  )
);
```

### Broadcast and Presence write

Authorize Presence and Broadcast write in one RLS policy.

```sql
CREATE POLICY "authenticated user can send Broadcast and Presence messages in a room"
ON "realtime"."messages"
FOR INSERT
TO authenticated
WITH CHECK (
  EXISTS (
    SELECT
      ru.user_id
    FROM
      rooms_users ru
    WHERE
      ru.user_id = auth.uid()
      AND ru.name = realtime.channel_name()

>       AND realtime.messages.extension in ['broadcast', 'presence']

Reviewer 2: ```suggestion
      AND realtime.messages.extension in ('broadcast', 'presence')
```

  )
);
```

## Avoid usage with Postgres Changes

Realtime Postgres Changes are separate from Channel authorization.

> Postgres Changes and Realtime Channl authorization are backwards compatible with your production application.

Reviewer 2: ```suggestion
Postgres Changes and Realtime Channel authorization are backwards compatible with your production application.
```
